File: java/caseys.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Caseys {

    /**
     * The state class represents an order state. The internal integer array
     * describes how much of each menu item was ordered.
     */
    public static class State implements Cloneable {
        public int[] counts;

        // hashCode has to be overridden for HashSets
        @Override
        public int hashCode() {
            // Use Arrays.hashCode()
            return Arrays.hashCode(counts);
        }

        // equals has to be overridden for HashSets
        @Override
        public boolean equals(Object other) {
            if (!(other instanceof State)) {
                return false;
            }
            State otherState = (State) other;
            // Use Arrays.equals()
            return Arrays.equals(otherState.counts, counts);
        }

        // clone is used later to duplicate the instance
        @Override
        public State clone() {
            State s = new State();
            // Use Arrays.copyOf()
            s.counts = Arrays.copyOf(counts, counts.length);
            return s;
        }
    }

    /**
     * This class represents a menu item, which has a name and a price.
     */
    public static class MenuItem {
        public String name;
        public int price;
    }

    // Array containing all available menu items
    public static MenuItem[] menu;
    // ArrayList containing all the solutions
    // They're all stored here so they can be printed later
    public static ArrayList<State> solutions = new ArrayList<State>();
    // Keep a Set of all the states that we've solved
    // This avoids duplicate cases and speeds up execution with Dynamic Programming
    public static Set<State> solved = new HashSet<>();

    // This a recursive function that generates the solutions
    public static void solve(State state, int money) {
        // Check if the state has already been solved
        if (solved.contains(state)) {
            return;
        }
        if (money == 0) {
            // If we come across a solution, make a copy of the state
            // And then add it to the solved ArrayList and Set
            State copy = state.clone();
            solved.add(copy);
            solutions.add(copy);
        } else {
            // If there's still money left go through every menu item
            for (int i = 0; i < menu.length; i++) {
                // If we can afford the item:
                if (menu[i].price <= money) {
                    // Update the state
                    state.counts[i]++;
                    // Recursively call solve with the updated state and new money
                    solve(state, money - menu[i].price);
                    // Reset the state
                    // This way we avoid expensive copy operations
                    state.counts[i]--;
                }
            }
        }
        // Finally mark the state as already visited for DP
        State copy = state.clone();
        solved.add(copy);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        // Read in goal and item count
        int goal = Integer.parseInt(in.readLine());
        int itemCount = Integer.parseInt(in.readLine());
        menu = new MenuItem[itemCount];
        // Read in all menu items
        for (int i = 0; i < itemCount; i++) {
            menu[i] = new MenuItem();
            String[] input = in.readLine().split(" ");
            menu[i].price = Integer.parseInt(input[1]);
            menu[i].name = input[0];
        }

        // Create initial state and call solve
        int[] stateArray = new int[itemCount];
        State state = new State();
        state.counts = stateArray;
        solve(state, goal);
        // If solution exists:
        if (solutions.size() > 0) {
            for (State solution : solutions) {
                System.out.println("------------------");
                for (int i = 0; i < solution.counts.length; i++) {
                    if (solution.counts[i] > 0) {
                        // Use printf to print out each item
                        System.out.printf("%s\tx%d\n", menu[i].name, solution.counts[i]);
                    }
                }
            }
            System.out.println("------------------");
        } else {
            // Otherwise mock the user
            System.out.println("How was your dinner?");
        }
    }
}

File: 2018-2019/py/caseys.py
# Read in money
money = int(input())
# Read in count
count = int(input())
# Read in all menu items
# Each menu item is a tuple of a name and a price
menu = []
for i in range(count):
    name, p = input().split()
    menu.append((name, int(p)))

# This array keeps track of all solutions to be printed later
solutions = []
# This set keeps track of which states have been visited
# This avoids duplicate states and speeds up execution (DP)
solved = set()
def solve(state, money):
    # Convert the state to a tuple so we can put it in a set
    state_t = tuple(state)
    # Check if we already came across this state
    if state_t in solved:
        return
    if money == 0:
        # If we have a solution, make a copy of the state and add it 
        # to the solutions list, and finally mark the current state as visited
        solution = state.copy()
        solutions.append(solution)
        solved.add(state_t)
    else:
        # If we still have money then iterate through every menu item
        for index, item in enumerate(menu):
            # If we can afford the item:
            if item[1] <= money:
                # Update the state
                state[index] += 1
                # Recursively call solve with the new state and price
                solve(state, money - item[1])
                # Reset the state
                # This way we don't need to make expensive copies
                state[index] -= 1
    # When everything is done mark the state as solved
    solved.add(state_t)

solve([0] * count, money)

# Print all solutions
if len(solutions) > 0:
    for solution in solutions:
        print("--------------------")
        for index, menu_item in enumerate(menu):
            # For every menu item in a solution
            # Print it if the count is greater than zero
            if solution[index] > 0:
                print(menu_item[0], "\tx", solution[index])
    print("--------------------")
else:
    print("How was your dinner?")

